File: lib/koto_cms/github.ex
defmodule KotoCms.GitHub do
  @base_url "https://api.github.com"

  def list_files(path \\ "", ref \\ nil) do
    repo = repo()
    branch = ref || default_branch()
    url = "#{@base_url}/repos/#{repo}/contents/#{path}?ref=#{branch}"

    case request(:get, url) do
      {:ok, data} when is_list(data) ->
        files = Enum.map(data, fn f ->
          %{
            path: f["path"],
            name: f["name"],
            type: if(f["type"] == "dir", do: "dir", else: "file"),
            sha: f["sha"]
          }
        end)
        |> Enum.sort_by(fn f -> {f.type != "dir", f.name} end)

        {:ok, files}

      error -> error
    end
  end

  def get_file(path, ref \\ nil) do
    repo = repo()
    branch = ref || default_branch()
    url = "#{@base_url}/repos/#{repo}/contents/#{path}?ref=#{branch}"

    case request(:get, url) do
      {:ok, data} ->
        content = data["content"] |> String.replace("\n", "") |> Base.decode64!()
        {:ok, %{path: data["path"], content: content, sha: data["sha"]}}

      error -> error
    end
  end

  def create_working_branch(fedi_handle) do
    repo = repo()
    branch_name = new_branch_name(fedi_handle)
    base_sha = get_latest_sha!(repo, default_branch())

    url = "#{@base_url}/repos/#{repo}/git/refs"
    body = %{ref: "refs/heads/#{branch_name}", sha: base_sha}

    case request(:post, url, body) do
      {:ok, _} -> {:ok, branch_name}
      error -> error
    end
  end

  def commit_file(path, content, sha, message, branch) do
    repo = repo()
    url = "#{@base_url}/repos/#{repo}/contents/#{path}"

    encoded = Base.encode64(content)
    body = %{message: message, content: encoded, branch: branch}
    body = if sha, do: Map.put(body, :sha, sha), else: body

    request(:put, url, body)
  end

  def delete_file(path, sha, message, branch) do
    repo = repo()
    url = "#{@base_url}/repos/#{repo}/contents/#{path}"
    body = %{message: message, sha: sha, branch: branch}

    request(:delete, url, body)
  end

  def list_user_prs(fedi_handle) do
    repo = repo()
    base = default_branch()
    slug = handle_slug(fedi_handle)
    prefix = "cms/#{slug}/"

    url = "#{@base_url}/repos/#{repo}/pulls?state=open&base=#{base}&per_page=50"

    with {:ok, prs} <- request(:get, url),
         {:ok, branches} <- get_branches(repo, prefix) do

      pr_map = Map.new(prs, fn pr -> {pr["head"]["ref"], pr} end)

      results = Enum.map(branches, fn branch ->
        pr = Map.get(pr_map, branch)
        %{
          branchName: branch,
          prNumber: pr && pr["number"],
          prUrl: pr && pr["html_url"],
          prState: cond do
            pr && pr["draft"] -> "draft"
            pr -> "open"
            true -> "none"
          end
        }
      end)
      |> Enum.sort_by(& &1.branchName, :desc)

      {:ok, results}
    end
  end

  def ensure_draft_pr(branch, fedi_handle) do
    case find_pr_for_branch(branch) do
      {:ok, pr} -> {:ok, %{prUrl: pr["html_url"], prNumber: pr["number"]}}
      {:error, :not_found} -> create_draft_pr(branch, fedi_handle)
    end
  end

  def mark_pr_ready(pr_number, title, body) do
    repo = repo()
    url = "#{@base_url}/repos/#{repo}/pulls/#{pr_number}"

    case request(:patch, url, %{title: title, body: body, draft: false}) do
      {:ok, data} -> {:ok, %{prUrl: data["html_url"], prNumber: data["number"]}}
      error -> error
    end
  end

  def get_preview_url(pr_number, config) do
    repo = repo()
    trusted_users = get_in(config, ["preview", "trustedUsers"]) || []
    url_patterns = get_in(config, ["preview", "urlPatterns"]) || []

    if Enum.empty?(trusted_users) or Enum.empty?(url_patterns) do
      {:ok, nil}
    else
      api_url = "#{@base_url}/repos/#{repo}/issues/#{pr_number}/comments?per_page=100"

      with {:ok, comments} <- request(:get, api_url) do
        preview_url =
          comments
          |> Enum.filter(fn comment ->
            Enum.any?(trusted_users, fn u -> u == comment["user"]["login"] end)
          end)
          |> Enum.flat_map(fn comment ->
            extract_urls(comment["body"] || "")
          end)
          |> Enum.find(fn url ->
            Enum.any?(url_patterns, fn pattern -> glob_match?(url, pattern) end)
          end)

        {:ok, preview_url}
      end
    end
  end

  def get_config(ref \\ nil) do
    case get_file(".koto.json", ref) do
      {:ok, %{content: content}} ->
        case Jason.decode(content) do
          {:ok, config} -> {:ok, config}
          _ -> {:ok, %{}}
        end
      _ -> {:ok, %{}}
    end
  end

  def list_tree(ref \\ nil) do
    repo = repo()
    branch = ref || default_branch()
    sha = get_latest_sha!(repo, branch)
    url = "#{@base_url}/repos/#{repo}/git/trees/#{sha}?recursive=1"

    case request(:get, url) do
      {:ok, %{"tree" => tree}} ->
        files =
          tree
          |> Enum.filter(fn item -> item["type"] == "blob" end)
          |> Enum.map(fn item ->
            name = item["path"] |> String.split("/") |> List.last()
            %{path: item["path"], name: name, type: "file", sha: item["sha"]}
          end)
        {:ok, files}
      error -> error
    end
  end

  def list_filtered_tree(config, ref \\ nil) do
    with {:ok, files} <- list_tree(ref) do
      include = Map.get(config, "include", [])
      exclude = Map.get(config, "exclude", [])
      include_patterns = if Enum.empty?(include), do: ["**"], else: include

      filtered =
        Enum.filter(files, fn f ->
          Enum.any?(include_patterns, &glob_match?(f.path, &1)) &&
            !Enum.any?(exclude, &glob_match?(f.path, &1))
        end)

      {:ok, filtered}
    end
  end

  def commit_message(action, path, author) do
    """
    content: #{action} #{path}

    Co-authored-by: #{author.name} <#{author.email}>
    Fediverse: `#{author.handle}`
    """
  end

  # Private

  defp request(method, url, body \\ nil) do
    headers = [
      {"Authorization", "Bearer #{token()}"},
      {"Accept", "application/vnd.github+json"},
      {"X-GitHub-Api-Version", "2022-11-28"},
      {"User-Agent", "koto-cms/1.0"}
    ]

    opts = [headers: headers]
    opts = if body, do: Keyword.put(opts, :json, body), else: opts

    case Req.request([method: method, url: url] ++ opts) do
      {:ok, %{status: status, body: body}} when status in 200..299 ->
        {:ok, body}

      {:ok, %{status: status, body: body}} ->
        {:error, "GitHub #{status}: #{inspect(body)}"}

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp get_latest_sha!(repo, branch) do
    url = "#{@base_url}/repos/#{repo}/git/refs/heads/#{branch}"
    {:ok, data} = request(:get, url)
    data["object"]["sha"]
  end

  defp find_pr_for_branch(branch) do
    repo = repo()
    base = default_branch()
    [owner, _] = String.split(repo, "/")

    url = "#{@base_url}/repos/#{repo}/pulls?state=open&head=#{owner}:#{branch}&base=#{base}"

    case request(:get, url) do
      {:ok, [pr | _]} -> {:ok, pr}
      {:ok, []} -> {:error, :not_found}
      error -> error
    end
  end

  defp create_draft_pr(branch, fedi_handle) do
    repo = repo()
    base = default_branch()
    date = Date.utc_today() |> Date.to_iso8601()

    url = "#{@base_url}/repos/#{repo}/pulls"
    body = %{
      title: "[Draft] CMS edit by `#{fedi_handle}` on #{date}",
      body: "Draft PR created by Koto.\n\nAuthor: `#{fedi_handle}`\nDate: #{date}",
      head: branch,
      base: base,
      draft: true
    }

    case request(:post, url, body) do
      {:ok, data} ->
        add_label(repo, data["number"], "cms")
        {:ok, %{prUrl: data["html_url"], prNumber: data["number"]}}

      error -> error
    end
  end

  defp add_label(repo, pr_number, label) do
    url = "#{@base_url}/repos/#{repo}/issues/#{pr_number}/labels"
    request(:post, url, %{labels: [label]})
  end

  defp get_branches(repo, prefix) do
    url = "#{@base_url}/repos/#{repo}/git/matching-refs/heads/#{prefix}"

    case request(:get, url) do
      {:ok, refs} ->
        branches = Enum.map(refs, fn ref ->
          String.replace(ref["ref"], "refs/heads/", "")
        end)
        {:ok, branches}

      error -> error
    end
  end

  defp handle_slug(fedi_handle) do
    fedi_handle
    |> String.replace(~r/^@/, "")
    |> String.replace("@", "-")
    |> String.replace(".", "-")
  end

  defp new_branch_name(fedi_handle) do
    slug = handle_slug(fedi_handle)
    date = Date.utc_today() |> Date.to_iso8601()
    rand = :crypto.strong_rand_bytes(3) |> Base.encode16(case: :lower)
    "cms/#{slug}/#{date}-#{rand}"
  end

  defp extract_urls(text) do
    Regex.scan(~r/https?:\/\/[^\s<>"'\)\]]+/, text)
    |> Enum.map(&List.first/1)
  end

  defp glob_match?(path, pattern) do
    regex_str =
      pattern
      |> String.split("**")
      |> Enum.map(fn part ->
        part
        |> String.split("*")
        |> Enum.map(&Regex.escape/1)
        |> Enum.join("[^/]*")
        |> String.replace("\\?", "[^/]")
      end)
      |> Enum.join(".*")

    case Regex.compile("^#{regex_str}$") do
      {:ok, regex} -> Regex.match?(regex, path)
      _ -> false
    end
  end

  defp token, do: System.get_env("GITHUB_BOT_TOKEN") || raise "GITHUB_BOT_TOKEN required"
  defp repo, do: System.get_env("GITHUB_REPO") || raise "GITHUB_REPO required"
  defp default_branch, do: System.get_env("GITHUB_BRANCH") || "main"
end

File: infrastructure/cloudflare-do/src/github.ts
import type {
  Author,
  GithubFileEntry,
  GithubFileResponse,
  KotoConfig,
  PrInfo,
  TenantConfig,
} from './types';
import { globMatch } from './glob';

const BASE = 'https://api.github.com';

export class GitHubError extends Error {
  status: number;
  constructor(message: string, status: number) {
    super(message);
    this.status = status;
  }
}

export class GitHub {
  constructor(private tenant: TenantConfig) {}

  private async request<T = unknown>(
    method: string,
    url: string,
    body?: unknown,
  ): Promise<T> {
    const headers: Record<string, string> = {
      Authorization: `Bearer ${this.tenant.githubToken}`,
      Accept: 'application/vnd.github+json',
      'X-GitHub-Api-Version': '2022-11-28',
      'User-Agent': 'koto-cms/1.0',
    };
    const init: RequestInit = { method, headers };
    if (body !== undefined) {
      headers['Content-Type'] = 'application/json';
      init.body = JSON.stringify(body);
    }
    const res = await fetch(url, init);
    const text = await res.text();
    const data = text ? JSON.parse(text) : null;
    if (!res.ok) {
      throw new GitHubError(`GitHub ${res.status}: ${text}`, res.status);
    }
    return data as T;
  }

  private get repo() {
    return this.tenant.repo;
  }

  private get baseBranch() {
    return this.tenant.defaultBranch;
  }

  async listFiles(path = '', ref?: string): Promise<GithubFileEntry[]> {
    const branch = ref || this.baseBranch;
    const url = `${BASE}/repos/${this.repo}/contents/${path}?ref=${branch}`;
    const data = await this.request<unknown>('GET', url);
    if (!Array.isArray(data)) return [];
    const files: GithubFileEntry[] = data.map((f) => {
      const entry = f as Record<string, unknown>;
      return {
        path: String(entry.path ?? ''),
        name: String(entry.name ?? ''),
        type: entry.type === 'dir' ? 'dir' : 'file',
        sha: String(entry.sha ?? ''),
      };
    });
    files.sort((a, b) => {
      const aDir = a.type === 'dir' ? 0 : 1;
      const bDir = b.type === 'dir' ? 0 : 1;
      if (aDir !== bDir) return aDir - bDir;
      return a.name.localeCompare(b.name);
    });
    return files;
  }

  async getFile(path: string, ref?: string): Promise<GithubFileResponse> {
    const branch = ref || this.baseBranch;
    const url = `${BASE}/repos/${this.repo}/contents/${path}?ref=${branch}`;
    const data = await this.request<Record<string, unknown>>('GET', url);
    const encoded = String(data.content ?? '').replace(/\n/g, '');
    const decoded = decodeBase64(encoded);
    return {
      path: String(data.path ?? path),
      content: decoded,
      sha: String(data.sha ?? ''),
    };
  }

  async getConfig(ref?: string): Promise<KotoConfig> {
    try {
      const file = await this.getFile('.koto.json', ref);
      try {
        return JSON.parse(file.content) as KotoConfig;
      } catch {
        return {};
      }
    } catch {
      return {};
    }
  }

  async listTree(ref?: string): Promise<GithubFileEntry[]> {
    const branch = ref || this.baseBranch;
    const sha = await this.getLatestSha(this.repo, branch);
    const url = `${BASE}/repos/${this.repo}/git/trees/${sha}?recursive=1`;
    const data = await this.request<{ tree?: Array<Record<string, unknown>> }>('GET', url);
    const out: GithubFileEntry[] = [];
    for (const entry of data.tree ?? []) {
      if (entry.type !== 'blob') continue;
      const path = String(entry.path ?? '');
      const name = path.split('/').pop() ?? '';
      out.push({ path, name, type: 'file', sha: String(entry.sha ?? '') });
    }
    return out;
  }

  async listFilteredTree(config: KotoConfig, ref?: string): Promise<GithubFileEntry[]> {
    const files = await this.listTree(ref);
    const include = config.include?.length ? config.include : ['**'];
    const exclude = config.exclude ?? [];
    return files.filter((f) => {
      const included = include.some((p) => globMatch(f.path, p));
      const excluded = exclude.some((p) => globMatch(f.path, p));
      return included && !excluded;
    });
  }

  async findOrCreateDraftBranch(fediHandle: string): Promise<string> {
    const prs = await this.listUserPrs(fediHandle);
    const draft = prs.find((p) => p.prState === 'draft');
    if (draft) return draft.branchName;
    return await this.createWorkingBranch(fediHandle);
  }

  async createWorkingBranch(fediHandle: string): Promise<string> {
    const branchName = newBranchName(fediHandle);
    const baseSha = await this.getLatestSha(this.repo, this.baseBranch);
    const url = `${BASE}/repos/${this.repo}/git/refs`;
    await this.request('POST', url, {
      ref: `refs/heads/${branchName}`,
      sha: baseSha,
    });
    return branchName;
  }

  async commitFile(
    path: string,
    content: string,
    sha: string | null,
    message: string,
    branch: string,
  ): Promise<unknown> {
    const url = `${BASE}/repos/${this.repo}/contents/${path}`;
    const body: Record<string, unknown> = {
      message,
      content: encodeBase64(content),
      branch,
    };
    if (sha) body.sha = sha;
    return await this.request('PUT', url, body);
  }

  /** Commit an already-base64-encoded blob (e.g. an uploaded image). */
  async commitBlob(
    path: string,
    contentBase64: string,
    sha: string | null,
    message: string,
    branch: string,
  ): Promise<Record<string, unknown>> {
    const url = `${BASE}/repos/${this.repo}/contents/${path}`;
    const body: Record<string, unknown> = { message, content: contentBase64, branch };
    if (sha) body.sha = sha;
    return await this.request<Record<string, unknown>>('PUT', url, body);
  }

  async deleteFile(
    path: string,
    sha: string,
    message: string,
    branch: string,
  ): Promise<unknown> {
    const url = `${BASE}/repos/${this.repo}/contents/${path}`;
    return await this.request('DELETE', url, { message, sha, branch });
  }

  async listUserPrs(fediHandle: string): Promise<PrInfo[]> {
    const base = this.baseBranch;
    const slug = handleSlug(fediHandle);
    const prefix = `cms/${slug}/`;

    const prsUrl = `${BASE}/repos/${this.repo}/pulls?state=open&base=${base}&per_page=50`;
    const prs = await this.request<Array<Record<string, unknown>>>('GET', prsUrl);
    const branches = await this.getBranches(prefix);

    const prByBranch = new Map<string, Record<string, unknown>>();
    for (const pr of prs) {
      const head = (pr.head as Record<string, unknown> | undefined)?.ref;
      if (typeof head === 'string') prByBranch.set(head, pr);
    }

    const results: PrInfo[] = branches.map((branch) => {
      const pr = prByBranch.get(branch);
      let state: PrInfo['prState'] = 'none';
      if (pr) state = pr.draft ? 'draft' : 'open';
      return {
        branchName: branch,
        prNumber: (pr?.number as number | undefined) ?? null,
        prUrl: (pr?.html_url as string | undefined) ?? null,
        prState: state,
      };
    });
    results.sort((a, b) => b.branchName.localeCompare(a.branchName));
    return results;
  }

  async ensureDraftPr(
    branch: string,
    fediHandle: string,
  ): Promise<{ prUrl: string; prNumber: number }> {
    const existing = await this.findPrForBranch(branch);
    if (existing) {
      return {
        prUrl: String(existing.html_url),
        prNumber: Number(existing.number),
      };
    }
    return await this.createDraftPr(branch, fediHandle);
  }

  async markPrReady(
    prNumber: number,
    title: string,
    body: string,
  ): Promise<{ prUrl: string; prNumber: number }> {
    const url = `${BASE}/repos/${this.repo}/pulls/${prNumber}`;
    const data = await this.request<Record<string, unknown>>('PATCH', url, {
      title,
      body,
      draft: false,
    });
    return {
      prUrl: String(data.html_url),
      prNumber: Number(data.number),
    };
  }

  async getPreviewUrl(prNumber: number, config: KotoConfig): Promise<string | null> {
    const trustedUsers = config.preview?.trustedUsers ?? [];
    const urlPatterns = config.preview?.urlPatterns ?? [];
    if (!trustedUsers.length || !urlPatterns.length) return null;

    const url = `${BASE}/repos/${this.repo}/issues/${prNumber}/comments?per_page=100`;
    const comments = await this.request<Array<Record<string, unknown>>>('GET', url);
    for (const comment of comments) {
      const user = comment.user as Record<string, unknown> | undefined;
      const login = user?.login as string | undefined;
      if (!login || !trustedUsers.includes(login)) continue;
      const body = (comment.body as string | undefined) ?? '';
      for (const u of extractUrls(body)) {
        if (urlPatterns.some((p) => globMatch(u, p))) return u;
      }
    }
    return null;
  }

  commitMessage(action: 'create' | 'update' | 'delete', path: string, author: Author): string {
    const subject = `${capitalize(action)} ${path}`;
    return [
      subject,
      '',
      'Edited via Koto CMS.',
      '',
      `Co-authored-by: ${author.name} <${author.email}>`,
      `Fediverse-handle: ${author.handle}`,
      '',
    ].join('\n');
  }

  // ── private ───────────────────────────────────────────────────

  private async getLatestSha(repo: string, branch: string): Promise<string> {
    const url = `${BASE}/repos/${repo}/git/refs/heads/${branch}`;
    const data = await this.request<{ object?: { sha?: string } }>('GET', url);
    return String(data.object?.sha ?? '');
  }

  private async findPrForBranch(branch: string): Promise<Record<string, unknown> | null> {
    const base = this.baseBranch;
    const [owner] = this.repo.split('/');
    const url = `${BASE}/repos/${this.repo}/pulls?state=open&head=${owner}:${branch}&base=${base}`;
    const prs = await this.request<Array<Record<string, unknown>>>('GET', url);
    return prs.length ? prs[0] : null;
  }

  private async createDraftPr(
    branch: string,
    fediHandle: string,
  ): Promise<{ prUrl: string; prNumber: number }> {
    const base = this.baseBranch;
    const date = new Date().toISOString().slice(0, 10);
    const url = `${BASE}/repos/${this.repo}/pulls`;
    const body = [
      'Content edits drafted via Koto CMS.',
      '',
      `Author: ${fediHandle}`,
      `Created: ${date}`,
    ].join('\n');
    const data = await this.request<Record<string, unknown>>('POST', url, {
      title: `Content edits by ${fediHandle}`,
      body,
      head: branch,
      base,
      draft: true,
    });
    const prNumber = Number(data.number);
    // Fire-and-forget label; ignore failure.
    try {
      await this.request('POST', `${BASE}/repos/${this.repo}/issues/${prNumber}/labels`, {
        labels: ['cms'],
      });
    } catch {
      // tolerate missing label permissions
    }
    return {
      prUrl: String(data.html_url),
      prNumber,
    };
  }

  private async getBranches(prefix: string): Promise<string[]> {
    const url = `${BASE}/repos/${this.repo}/git/matching-refs/heads/${prefix}`;
    const refs = await this.request<Array<Record<string, unknown>>>('GET', url);
    return refs.map((r) => String(r.ref ?? '').replace('refs/heads/', ''));
  }
}

function capitalize(s: string): string {
  return s.length ? s[0].toUpperCase() + s.slice(1) : s;
}

function handleSlug(handle: string): string {
  return handle.replace(/^@/, '').replace(/@/g, '-').replace(/\./g, '-');
}

function newBranchName(fediHandle: string): string {
  const slug = handleSlug(fediHandle);
  const date = new Date().toISOString().slice(0, 10);
  const rand = Array.from(crypto.getRandomValues(new Uint8Array(3)))
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
  return `cms/${slug}/${date}-${rand}`;
}

function extractUrls(text: string): string[] {
  const re = /https?:\/\/[^\s<>"'\)\]]+/g;
  return Array.from(text.matchAll(re)).map((m) => m[0]);
}

function encodeBase64(input: string): string {
  const bytes = new TextEncoder().encode(input);
  let bin = '';
  for (const b of bytes) bin += String.fromCharCode(b);
  return btoa(bin);
}

function decodeBase64(input: string): string {
  const bin = atob(input);
  const bytes = new Uint8Array(bin.length);
  for (let i = 0; i < bin.length; i++) bytes[i] = bin.charCodeAt(i);
  return new TextDecoder().decode(bytes);
}
